### PR TITLE
Add logdet and logabsdet methods for Symmetric/Hermitian matrices

### PR DESCRIPTION
**Pull request by ElOceanografo**
There was a method for `det` already, but not for `log[abs]det`.

### CODE REVIEW

File: stdlib/LinearAlgebra/src/symmetric.jl [hunk 1]
```diff
@@ -574,9 +574,13 @@ function _factorize(A::HermOrSym{T}; check::Bool=true) where T
     end
 end
 
-det(A::RealHermSymComplexHerm) = real(det(_factorize(A; check=false)))
-det(A::Symmetric{<:Real}) = det(_factorize(A; check=false))
-det(A::Symmetric) = det(_factorize(A; check=false))
+for f in (:logabsdet, :logdet, :det)
+    @eval begin
+        $f(A::RealHermSymComplexHerm) = real($f(_factorize(A; check=false)))
+        $f(A::Symmetric{<:Real}) = $f(_factorize(A; check=false))
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/symmetric.jl:580]: This one is included in the one above, isn't it?

**Review by dkarrasch — CHANGES_REQUESTED** [review]
This needs tests. In particular, `logabsdet` returns a tuple, on which `real` is not defined. Another question is whether these methods are needed at all. Don't they fall back on the `det` implementation? It seems that all of these three methods are working, e.g., for some real `Symmetric` matrix.

**Comment by ElOceanografo** [comment]
The `logdet` methods do not in general fall back on `det`, rather they do a `lu` decomposition of the matrix and call `logabsdet` on that. These methods are needed in particular for the case where `Symmetric` or `Hermitian` is wrapping a sparse matrix. Currently in this situation, `logabsdet` ends up calling a [generic version](https://github.com/JuliaLang/julia/blob/bed2cd540a11544ed4be381d471bbf590f0b745e/stdlib/LinearAlgebra/src/lu.jl#L133) of `lu` which is orders of magnitude slower than the specialized sparse one in UMFPACK:
```julia
using LinearAlgebra, SparseArrays
Q = sprandn(1000, 1000, 0.0002)
Q = Q'Q + I
@time logdet(Q) # 0.000180 seconds (108 allocations: 559.406 KiB)
@time logdet(Symmetric(Q)) # 5.488019 seconds (7 allocations: 82.953 KiB)
```

This PR seemed like the minimal change to allow the factorization method to specialize on the underlying matrix structure, though I'm open to being corrected on that!

**Comment by dkarrasch** [comment]
Ok, the failing tests indicate that these methods are exercised. It remains to fix the `logabsdet` definition, perhaps using broadcasting?

File: stdlib/LinearAlgebra/src/symmetric.jl (hunk 1 above)
Comment (ElOceanografo) [reply stdlib/LinearAlgebra/src/symmetric.jl]: I guess so? I found it like that so didn't change it.

File: stdlib/LinearAlgebra/src/symmetric.jl [hunk 2]
```diff
@@ -574,6 +574,12 @@ function _factorize(A::HermOrSym{T}; check::Bool=true) where T
     end
 end
 
+logabsdet(A::RealHermSymComplexHerm) = logabsdet(_factorize(A; check=false))
+logabsdet(A::Symmetric{<:Real}) = logabsdet(_factorize(A; check=false))
+logabsdet(A::Symmetric) = logabsdet(_factorize(A; check=false))
+logdet(A::RealHermSymComplexHerm) = real(logdet(_factorize(A; check=false)))
+logdet(A::Symmetric{<:Real}) = logdet(_factorize(A; check=false))
+logdet(A::Symmetric) = logdet(_factorize(A; check=false))
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/symmetric.jl:582]: `Symmetric{<:Real}` is included in `RealHermSymComplexHerm`, so if you want the same definition to work, then there's no need to have that extra method.

```suggestion
logabsdet(A::RealHermSymComplexHerm) = logabsdet(_factorize(A; check=false))
logabsdet(A::Symmetric) = logabsdet(_factorize(A; check=false))
logdet(A::RealHermSymComplexHerm) = real(logdet(_factorize(A; check=false)))
logdet(A::Symmetric) = logdet(_factorize(A; check=false))
```

Comment (ElOceanografo) [reply stdlib/LinearAlgebra/src/symmetric.jl]: It looks like removing these definitions has introduced a method signature ambiguity:
```
  MethodError: logdet(::LinearAlgebra.Symmetric{Float32, Matrix{Float32}}) is ambiguous.
  Candidates:
    logdet(A::LinearAlgebra.Symmetric)
      @ LinearAlgebra /cache/build/default-amdci4-7/julialang/julia-master/julia-e9c7325a82/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:580
    logdet(A::Union{LinearAlgebra.Hermitian{T, S}, LinearAlgebra.Hermitian{Complex{T}, S}, LinearAlgebra.Symmetric{T, S}} where {T<:Real, S})
      @ LinearAlgebra /cache/build/default-amdci4-7/julialang/julia-master/julia-e9c7325a82/share/julia/stdlib/v1.11/LinearAlgebra/src/symmetric.jl:579
  Possible fix, define
    logdet(::LinearAlgebra.Symmetric{T, S} where {T<:Real, S<:(AbstractMatrix{<:T})})

```
Going to add them back in.

**Review by dkarrasch — COMMENTED** [review]
I think we should have specific tests for these cases, both real and complex, both positive and negative sign, and nail down the desired behaviour.

File: stdlib/LinearAlgebra/src/symmetric.jl [hunk 3]
```diff
@@ -574,6 +574,12 @@ function _factorize(A::HermOrSym{T}; check::Bool=true) where T
     end
 end
 
+logabsdet(A::RealHermSymComplexHerm) = ((l, s) = logabsdet(_factorize(A; check=false)); return real(l), s)
+logabsdet(A::Symmetric{<:Real}) = logabsdet(_factorize(A; check=false))
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/symmetric.jl:578]: The same `real`ification should then probably apply here and in lines 581 and 584, shouldn't it? Or are they guaranteed to yield real return types?

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetric.jl:578]: Seems like this is guaranteed to be real already.

**Review by dkarrasch — APPROVED** [review]

**Pull request merged**